presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -214,6 +216,11 @@ else if (exchange.getType() == ExchangeNode.Type.REPLICATE) {
                     .map(PlanFragment::getId)
                     .collect(toImmutableList());
 
+            if (exchange.getType() == ExchangeNode.Type.MERGE_GATHER) {
+                checkState(exchange.getSources().get(0) instanceof SortNode, "Merge must be preceded by sort");
```

Comment:
I'll add an assertion there's always a single source, but we should consider adding a `MergeExchangeNode` and storing the `orderBy` and `orderings` there, instead of getting those from the source node. I didn't do that in the initial PR because it adds a lot of code and complication, but I can add it.